### PR TITLE
Partial support for filesystems in FileSet

### LINKED ISSUE #373: Feature request: Add fsspec FileSystem support in FileSet
The [fsspec library](https://filesystem-spec.readthedocs.io) provides a generic interface to interact with filesystems independently of where they are located.  This means users can automatically search for files on remote filesystems such as with [the s3fs implementation](https://s3fs.readthedocs.io/en/latest/) for Amazon S3 buckets, or wrap this in a [`CachingFileSystem` instance](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.implementations.cached.CachingFileSystem) caching requested data locally in sparse files.  For users who don't have all satellite data available locally the ability to download only the data needed can save significant time.

My specific use case is to find [GLM LCFA](https://docs.opendata.aws/noaa-goes16/cics-readme.html) files covering a specified time period, pass those to [glmtools](https://github.com/deeplycloudy/glmtools) (ideally by passing files opened with a [`CachingFileSystem` instance](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.implementations.cached.CachingFileSystem), but passing files downloaded entirely if needed), so that I can read the resulting gridded products using [satpy](https://satpy.readthedocs.io).

Satpy contains functionality to search for files based on file patterns, similar to what exists in a [typhon Fileset](https://radiativetransfer.org/misc/typhon/doc-trunk/tutorials/fileset.html), and this functionality supports fsspec FileSystem instances; however, it only supports sensors for which Satpy has a reading routine (Satpy reads gridded GLM, but not ungridded GLM LCFA).  It also has [ongoing work](https://github.com/pytroll/satpy/pull/1321) for generic filesystem support, but this will not help for GLM [as long as glmtools interfacing happens outside of satpy](https://github.com/pytroll/satpy/issues/1386) (and the latter does not support fsspec filesystems).  

Therefore, it would be very nice if the user could pass any instance implementing [`fsspec.spec.AbstractFileSystem`](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.spec.AbstractFileSystem), so that the `FileSet` functionality is available regardless of where the files are located.

### PR DESCRIPTION
**Pull request by gerritholl**
This PR adds partial support for fsspec AbstractFileSystem implementation in FileSet, in particular for finding files.  An example using `sf3s`:

```python
import s3fs
from typhon.files.fileset import FileSet
abi_fileset = FileSet(
        path="noaa-goes16/ABI-L1b-RadF/{year}/{doy}/{hour}/OR_ABI-L1b-RadF-M6C02_G16_s{year}{doy}{hour}{minute}{second}*_e{end_year}{end_doy}{end_hour}{end_minute}{end_second}*_c*.nc",                                                                                                 
        name="abi",
        fs=s3fs.S3FileSystem(anon=True))
for f in abi_fileset.find("2019-11-18T05:30", "2019-11-18T06:15"):
    print(f)
```

Gives:

```
noaa-goes16/ABI-L1b-RadF/2019/322/05/OR_ABI-L1b-RadF-M6C02_G16_s20193220530285_e20193220539593_c20193220540030.nc
noaa-goes16/ABI-L1b-RadF/2019/322/05/OR_ABI-L1b-RadF-M6C02_G16_s20193220540285_e20193220549593_c20193220550041.nc
noaa-goes16/ABI-L1b-RadF/2019/322/05/OR_ABI-L1b-RadF-M6C02_G16_s20193220550285_e20193220559593_c20193220600041.nc
noaa-goes16/ABI-L1b-RadF/2019/322/06/OR_ABI-L1b-RadF-M6C02_G16_s20193220600285_e20193220609593_c20193220610030.nc
noaa-goes16/ABI-L1b-RadF/2019/322/06/OR_ABI-L1b-RadF-M6C02_G16_s20193220610285_e20193220619593_c20193220620040.nc
```

- [x] Closes #373 
- [x] Tests added
- [ ] Tests passed
- [x] Documentation added

### CODE REVIEW

**Comment by gerritholl** [comment]
I don't know why it still fails on Windows despite replacing / by os.sep.

**Comment by gerritholl** [comment]
This still fails on Windows (I don't know why) and it doesn't work as intended with S3FileSystem (it seems globbing behaves differently there): `fs_s3.glob("noaa-goes16/ABI-L1b-RadF/2020/045/*")` returns nothing even though `fs_s3.ls("noaa-goes16/ABI-L1b-RadF/2020/045/")` does, and `fs_s3.glob("noaa-goes16/ABI-L1b-RadF/2020/*")` is very slow, suggesting there is something recursive going on.

**Comment by gerritholl** [comment]
Finding files is extremely slow due to globbing in s3fs being extremely slow, probably due to the problem reported at https://github.com/dask/s3fs/issues/378.

**Comment by gerritholl** [comment]
Note that at present the s3fs example only works with s3fs 0.5.0 and not with the latest released s3fs 0.5.1, due to the bug reported at https://github.com/dask/s3fs/issues/378.  Hopefully that problem will be fixed so that this searching works with s3fs 0.5.2 or 0.6.

**Comment by gerritholl** [comment]
Can anyone with access to a Windows Python installation shed light on why the tests may be failing on Windows?  Did the tests actually succeed on Windows prior to this PR?

**Comment by olemke** [comment]
Hi @gerritholl! I can test it on a Windows machine, but not before some time later next week since I don't have a setup ready.
However, the tests were working before this PR as you can see in the build log: https://github.com/atmtools/typhon/actions/runs/288892462

**Comment by gerritholl** [comment]
Meanwhile https://github.com/dask/s3fs/pull/379 was merged so the typhon functionality and the documented example should work with the latest s3fs master or the next release.

**Comment by olemke** [comment]
This seems to be caused by inconsistent handling of path separators. One place I found to be inconsistent is in

https://github.com/gerritholl/typhon/blob/af22c48f0721e2961388a9305cc1cde26a7cf6a1/typhon/files/fileset.py#L1328-L1329

`self.file_system.glob` returns a path with `/` separators while `regex` uses the `\\` version. There might be other places where this causes a problem, but this is all I had time to look at now.

**Comment by gerritholl** [comment]
Ah, that makes sense.  If we search on a remote file system like this PR intends to support, then the assumption that the file system uses `os.sep` as a separator is no longer guaranteed to be true.  I will try to think of a solution.

**Comment by gerritholl** [comment]
With all the commits to make it work on Windows it has become regrettably somewhat ugly, and perhaps breaking backward compatibility despite tests passing.  Can Windows users please comment if it breaks their workflow or not?

**Comment by olemke** [comment]
I don't know if there are any Windows users of Typhon at the moment, so don't expect an answer to your question. We mainly added Windows as a supported platform because Typhon is a Python-only package and it seemed like the right thing to do to be good Python citizens.
Thanks for putting in the work to keep this cross-platform. From my side this is ready to be merged. I'll have a chat with Lukas if he's also fine with it when he's back next week. I don't think you'll need to add John to PRs in the future since he has moved on to other things.

**Review by olemke — APPROVED** [review]

File: typhon/files/handlers/common.py [hunk 1]
```diff
@@ -253,6 +255,12 @@ def write(self, data, filename, **kwargs):
             "This file handler does not support writing data to a file. You "
             "should use a different file handler.")
 
+    def _ensure_local_filesystem(self, file_info):
+        if not isinstance(file_info.file_system, LocalFileSystem):
+            raise NotImplementedError(
+                    f"File handler {str(type(self).__name__):s} can only "
```
Comment (lkluft) [inline typhon/files/handlers/common.py]: I guess ```type(...).__name__``` is always a string.

File: typhon/files/handlers/common.py [hunk 2]
```diff
@@ -319,7 +330,8 @@ def __hash__(self):
     def __repr__(self):
         return f"FileInfo(\n  '{self.path}',\n" \
                f"  times={self.times},\n" \
-               f"  attr={self.attr}\n)"
+               f"  attr={self.attr},\n" \
+               f"  fs={self.file_system})"
```
Comment (lkluft) [inline typhon/files/handlers/common.py:334]: The ``__repr__`` method is used to represent an object in containers (e.g. a list of `FileInfo`) so this implementation might be a bit wordy. But this is just a fine-tuning comment that you can happily ignore.

**Review by lkluft — APPROVED** [review]
Looks good in general 👍  I only left some minor comments that you may address or ignore ;)

File: typhon/tests/files/test_fileset.py [hunk 3]
```diff
@@ -12,32 +15,76 @@ class TestFileSet:
...
+            fn = cont[0]
+            if fn[0] not in (os.sep, "/", self.refdir[0]):
+                return self.refdir[1:]
+        except (IndexError, AttributeError, StopIteration, KeyError, TypeError):
+            pass
+        return self.refdir
+
+    def init_filesets(self, fs=None):
+#        if self.filesets is not None:
+#            return self.filesets
```
[the remark this answers is not in the document]
Comment (gerritholl) [reply typhon/tests/files/test_fileset.py]: Fixed

File: typhon/files/handlers/common.py (hunk 1 above)
Comment (gerritholl) [reply typhon/files/handlers/common.py]: True, simplified.

**Comment by gerritholl** [comment]
I thought I had write access here, but apparently not (anymore?) as no "merge" button shows up ("Only those with write access to this repository can merge pull requests.).

**Comment by olemke** [comment]
Merges can only be done by maintainers which are currently Lukas and me.
